changelog.d/12319.bugfix

```diff
@@ -0,0 +1 @@
+Fix bug with incremental sync missing events when rejoining/backfilling. Contributed by Nick @ Beeper.
```

Comment:
As a note, I'm not a Synapse reviewer/maintainer. After we get the tests passing again ✅, we should assign this back to `matrix-org/synapse-core`